File: clients/python/uiform/types/documents/extractions.py
import base64
import datetime
from typing import Any, Literal, Optional

from anthropic.types.message import Message
from anthropic.types.message_param import MessageParam
from openai.types.chat import ChatCompletionMessageParam
from openai.types.chat.chat_completion import ChatCompletion
from openai.types.chat.chat_completion_chunk import ChatCompletionChunk
from openai.types.chat.chat_completion_chunk import Choice as ChoiceChunk
from openai.types.chat.chat_completion_chunk import ChoiceDelta as ChoiceDeltaChunk
from openai.types.chat.chat_completion_reasoning_effort import ChatCompletionReasoningEffort
from openai.types.chat.parsed_chat_completion import ParsedChatCompletion, ParsedChoice
from openai.types.responses.response import Response
from openai.types.responses.response_input_param import ResponseInputItemParam
from pydantic import BaseModel, ConfigDict, Field, ValidationInfo, computed_field, field_validator, model_validator

from ..._utils.usage.usage import compute_cost_from_model, compute_cost_from_model_with_breakdown, CostBreakdown

from ..._utils.ai_models import find_provider_from_model
from ..ai_models import AIProvider, Amount, get_model_card
from ..chat import ChatCompletionUiformMessage
from ..mime import MIMEData
from ..modalities import Modality
from ..standards import ErrorDetail, StreamingBaseModel


class DocumentExtractRequest(BaseModel):
    model_config = ConfigDict(arbitrary_types_allowed=True)

    document: MIMEData = Field(..., description="Document to be analyzed")
    modality: Modality
    image_resolution_dpi: int = Field(default=96, description="Resolution of the image sent to the LLM")
    browser_canvas: Literal['A3', 'A4', 'A5'] = Field(default='A4', description="Sets the size of the browser canvas for rendering documents in browser-based processing. Choose a size that matches the document type.")
    model: str = Field(..., description="Model used for chat completion")
    json_schema: dict[str, Any] = Field(..., description="JSON schema format used to validate the output data.")
    temperature: float = Field(default=0.0, description="Temperature for sampling. If not provided, the default temperature for the model will be used.", examples=[0.0])
    reasoning_effort: ChatCompletionReasoningEffort = Field(
        default="medium", description="The effort level for the model to reason about the input data. If not provided, the default reasoning effort for the model will be used."
    )
    n_consensus: int = Field(default=1, description="Number of consensus models to use for extraction. If greater than 1 the temperature cannot be 0.")
    # Regular fields
    stream: bool = Field(default=False, description="If true, the extraction will be streamed to the user using the active WebSocket connection")
    seed: int | None = Field(default=None, description="Seed for the random number generator. If not provided, a random seed will be generated.", examples=[None])
    store: bool = Field(default=True, description="If true, the extraction will be stored in the database")
    need_validation: bool = Field(default=False, description="If true, the extraction will be validated against the schema")

    # Add a model validator that rejects n_consensus > 1 if temperature is 0
    @field_validator("n_consensus")
    def check_n_consensus(cls, v: int, info: ValidationInfo) -> int:
        if v > 1 and info.data.get("temperature") == 0:
            raise ValueError("n_consensus greater than 1 but temperature is 0")
        return v


class ConsensusModel(BaseModel):
    model: str = Field(description="Model name")
    temperature: float = Field(default=0.0, description="Temperature for consensus")
    reasoning_effort: ChatCompletionReasoningEffort = Field(
        default="medium", description="The effort level for the model to reason about the input data. If not provided, the default reasoning effort for the model will be used."
    )



# For location of fields in the document (OCR)
class FieldLocation(BaseModel):
    label: str = Field(..., description="The label of the field")
    value: str = Field(..., description="The extracted value of the field")
    quote: str = Field(..., description="The quote of the field (verbatim from the document)")
    file_id: str | None = Field(default=None, description="The ID of the file")
    page: int | None = Field(default=None, description="The page number of the field (1-indexed)")
    bboxes_normalized: list[tuple[float, float, float, float]] | None = Field(default=None, description="The normalized bounding boxes of the field")
    score: float | None = Field(default=None, description="The score of the field")
    match_level: Literal["token", "line", "block"] | None = Field(default=None, description="The level of the match (token, line, block)")


class UiParsedChoice(ParsedChoice):
    # Adaptable ParsedChoice that allows None for the finish_reason
    finish_reason: Literal["stop", "length", "tool_calls", "content_filter", "function_call"] | None = None  # type: ignore
    field_locations: dict[str, list[FieldLocation]] | None = Field(default=None, description="The locations of the fields in the document, if available")
    key_mapping: dict[str, Optional[str]] | None = Field(default=None, description="Mapping of consensus keys to original model keys")


LikelihoodsSource = Literal["consensus", "log_probs"]


class UiParsedChatCompletion(ParsedChatCompletion):
    extraction_id: str | None = None
    choices: list[UiParsedChoice]
    # Additional metadata fields (UIForm)
    likelihoods: Optional[dict[str, Any]] = Field(
        default=None, description="Object defining the uncertainties of the fields extracted when using consensus. Follows the same structure as the extraction object."
    )
    schema_validation_error: ErrorDetail | None = None
    # Timestamps
    request_at: datetime.datetime | None = Field(default=None, description="Timestamp of the request")
    first_token_at: datetime.datetime | None = Field(default=None, description="Timestamp of the first token of the document. If non-streaming, set to last_token_at")
    last_token_at: datetime.datetime | None = Field(default=None, description="Timestamp of the last token of the document")

    @computed_field
    @property
    def api_cost(self) -> Optional[Amount]:
        if self.usage:
            try:
                cost = compute_cost_from_model(self.model, self.usage)
                return cost
            except Exception as e:
                print(f"Error computing cost: {e}")
                return None
        return None


class UiResponse(Response):
    extraction_id: str | None = None
    # Additional metadata fields (UIForm)
    likelihoods: Optional[dict[str, Any]] = Field(
        default=None, description="Object defining the uncertainties of the fields extracted when using consensus. Follows the same structure as the extraction object."
    )
    schema_validation_error: ErrorDetail | None = None
    # Timestamps
    request_at: datetime.datetime | None = Field(default=None, description="Timestamp of the request")
    first_token_at: datetime.datetime | None = Field(default=None, description="Timestamp of the first token of the document. If non-streaming, set to last_token_at")
    last_token_at: datetime.datetime | None = Field(default=None, description="Timestamp of the last token of the document")


class LogExtractionRequest(BaseModel):
    messages: list[ChatCompletionUiformMessage] | None = None  # TODO: compatibility with Anthropic
    openai_messages: list[ChatCompletionMessageParam] | None = None
    openai_responses_input: list[ResponseInputItemParam] | None = None
    anthropic_messages: list[MessageParam] | None = None
    anthropic_system_prompt: str | None = None
    document: MIMEData = Field(
        default=MIMEData(
            filename="dummy.txt",
            # url is a base64 encoded string with the mime type and the content. For the dummy one we will send a .txt file with the text "No document provided"
            url="data:text/plain;base64," + base64.b64encode(b"No document provided").decode("utf-8"),
        ),
        description="Document analyzed, if not provided a dummy one will be created with the text 'No document provided'",
    )
    completion: dict | UiParsedChatCompletion | Message | ParsedChatCompletion | ChatCompletion | None = None
    openai_responses_output: Response | None = None
    json_schema: dict[str, Any]
    model: str
    temperature: float

    # Validate that at least one of the messages, openai_messages, anthropic_messages is provided using model_validator
    @model_validator(mode="before")
    def validation(cls, data: Any) -> Any:
        messages_candidates = [data.get("messages"), data.get("openai_messages"), data.get("anthropic_messages"), data.get("openai_responses_input")]
        messages_candidates = [candidate for candidate in messages_candidates if candidate is not None]
        if len(messages_candidates) != 1:
            raise ValueError("Exactly one of the messages, openai_messages, anthropic_messages, openai_responses_input must be provided")

        # Validate that if anthropic_messages is provided, anthropic_system_prompt is also provided
        if data.get("anthropic_messages") is not None and data.get("anthropic_system_prompt") is None:
            raise ValueError("anthropic_system_prompt must be provided if anthropic_messages is provided")

        completion_candidates = [data.get("completion"), data.get("openai_responses_output")]
        completion_candidates = [candidate for candidate in completion_candidates if candidate is not None]
        if len(completion_candidates) != 1:
            raise ValueError("Exactly one of completion, openai_responses_output must be provided")

        return data


class LogExtractionResponse(BaseModel):
    extraction_id: str | None = None  # None only in case of error
    status: Literal["success", "error"]
    error_message: str | None = None


# DocumentExtractResponse = UiParsedChatCompletion


###### I'll place here for now -- New Streaming API


# We build from the openai.types.chat.chat_completion_chunk.ChatCompletionChunk adding just two three additional fields:
# - missing_content: list[str]              #  The missing content (to be concatenated to the cumulated content to create a potentially valid JSON), it should be aligned with the choices index
# - is_valid_json: list[bool]               #  Whether the total accumulated content is a valid JSON
# - likelihoods: dict[str, float]     #  The delta of the flattened likelihoods (to be merged with the cumulated likelihoods)
# - schema_validation_error: ErrorDetail | None = None #  The error in the schema validation of the total accumulated content


class UiParsedChoiceDeltaChunk(ChoiceDeltaChunk):
    flat_likelihoods: dict[str, float] = {}
    flat_parsed: dict[str, Any] = {}
    flat_deleted_keys: list[str] = []
    field_locations: dict[str, list[FieldLocation]] | None = Field(default=None, description="The locations of the fields in the document, if available")
    missing_content: str = ""
    is_valid_json: bool = False
    key_mapping: dict[str, Optional[str]] | None = Field(default=None, description="Mapping of consensus keys to original model keys")


class UiParsedChoiceChunk(ChoiceChunk):
    delta: UiParsedChoiceDeltaChunk


class UiParsedChatCompletionChunk(StreamingBaseModel, ChatCompletionChunk):
    extraction_id: str | None = None
    choices: list[UiParsedChoiceChunk]
    schema_validation_error: ErrorDetail | None = None
    # Timestamps
    request_at: datetime.datetime | None = Field(default=None, description="Timestamp of the request")
    first_token_at: datetime.datetime | None = Field(default=None, description="Timestamp of the first token of the document. If non-streaming, set to last_token_at")
    last_token_at: datetime.datetime | None = Field(default=None, description="Timestamp of the last token of the document")

    @computed_field
    @property
    def api_cost(self) -> Optional[Amount]:
        if self.usage:
            try:
                cost = compute_cost_from_model(self.model, self.usage)
                return cost
            except Exception as e:
                print(f"Error computing cost: {e}")
                return None
        return None

    @computed_field  # type: ignore
    @property
    def cost_breakdown(self) -> Optional[CostBreakdown]:
        if self.usage:
            try:
                cost = compute_cost_from_model_with_breakdown(self.model, self.usage)
                return cost
            except Exception as e:
                print(f"Error computing cost: {e}")
                return None
        return None

    def chunk_accumulator(self, previous_cumulated_chunk: "UiParsedChatCompletionChunk | None" = None) -> "UiParsedChatCompletionChunk":
        """
        Accumulate the chunk into the state, returning a new UiParsedChatCompletionChunk with the accumulated content that could be yielded alone to generate the same state.
        """

        def safe_get_delta(chnk: "UiParsedChatCompletionChunk | None", index: int) -> UiParsedChoiceDeltaChunk:
            if chnk is not None and index < len(chnk.choices):
                return chnk.choices[index].delta
            else:
                return UiParsedChoiceDeltaChunk(
                    content="",
                    flat_parsed={},
                    flat_likelihoods={},
                    missing_content="",
                    is_valid_json=False,
                )

        max_choices = max(len(self.choices), len(previous_cumulated_chunk.choices)) if previous_cumulated_chunk is not None else len(self.choices)

        # Get the current chunk missing content, flat_deleted_keys and is_valid_json
        acc_missing_content = [safe_get_delta(self, i).missing_content or "" for i in range(max_choices)]
        acc_flat_deleted_keys = [safe_get_delta(self, i).flat_deleted_keys for i in range(max_choices)]
        acc_is_valid_json = [safe_get_delta(self, i).is_valid_json for i in range(max_choices)]
        acc_field_locations = [safe_get_delta(self, i).field_locations for i in range(max_choices)]  # This is only present in the last chunk.
        # Delete from previous_cumulated_chunk.choices[i].delta.flat_parsed the keys that are in safe_get_delta(self, i).flat_deleted_keys
        for i in range(max_choices):
            previous_delta = safe_get_delta(previous_cumulated_chunk, i)
            current_delta = safe_get_delta(self, i)
            for deleted_key in current_delta.flat_deleted_keys:
                previous_delta.flat_parsed.pop(deleted_key, None)
                previous_delta.flat_likelihoods.pop(deleted_key, None)
        # Accumulate the flat_parsed and flat_likelihoods
        acc_flat_parsed = [safe_get_delta(previous_cumulated_chunk, i).flat_parsed | safe_get_delta(self, i).flat_parsed for i in range(max_choices)]
        acc_flat_likelihoods = [safe_get_delta(previous_cumulated_chunk, i).flat_likelihoods | safe_get_delta(self, i).flat_likelihoods for i in range(max_choices)]
        acc_key_mapping = [safe_get_delta(previous_cumulated_chunk, i).key_mapping or safe_get_delta(self, i).key_mapping for i in range(max_choices)]

        acc_content = [(safe_get_delta(previous_cumulated_chunk, i).content or "") + (safe_get_delta(self, i).content or "") for i in range(max_choices)]
        usage = self.usage
        first_token_at = self.first_token_at
        last_token_at = self.last_token_at
        request_at = self.request_at

        return UiParsedChatCompletionChunk(
            extraction_id=self.extraction_id,
            id=self.id,
            created=self.created,
            model=self.model,
            object=self.object,
            usage=usage,
            choices=[
                UiParsedChoiceChunk(
                    delta=UiParsedChoiceDeltaChunk(
                        content=acc_content[i],
                        flat_parsed=acc_flat_parsed[i],
                        flat_likelihoods=acc_flat_likelihoods[i],
                        flat_deleted_keys=acc_flat_deleted_keys[i],
                        field_locations=acc_field_locations[i],
                        missing_content=acc_missing_content[i],
                        is_valid_json=acc_is_valid_json[i],
                        key_mapping=acc_key_mapping[i],
                    ),
                    index=i,
                )
                for i in range(max_choices)
            ],
            schema_validation_error=self.schema_validation_error,
            request_at=request_at,
            first_token_at=first_token_at,
            last_token_at=last_token_at,
        )

File: clients/node/src/types/documents/extractions.ts
import { z } from 'zod';
import { MIMEDataSchema } from '../mime.js';
import { ModalitySchema } from '../modalities.js';
import { BrowserCanvasSchema } from '../browser_canvas.js';
import { ChatCompletionRetabMessageSchema } from '../chat.js';
import { ErrorDetail } from '../standards.js';

// Reasoning effort type
export const ChatCompletionReasoningEffortSchema = z.enum(['low', 'medium', 'high']);
export type ChatCompletionReasoningEffort = z.infer<typeof ChatCompletionReasoningEffortSchema>;

// Document Extract Request
export const DocumentExtractRequestSchema = z.object({
  document: MIMEDataSchema.optional(),
  documents: z.array(MIMEDataSchema),
  modality: ModalitySchema,
  image_resolution_dpi: z.number().int().default(96),
  browser_canvas: BrowserCanvasSchema.default('A4'),
  model: z.string(),
  json_schema: z.record(z.any()),
  temperature: z.number().default(0.0),
  reasoning_effort: ChatCompletionReasoningEffortSchema.default('medium'),
  n_consensus: z.number().int().default(1),
  stream: z.boolean().default(false),
  seed: z.number().int().nullable().optional(),
  store: z.boolean().default(true),
  need_validation: z.boolean().default(false),
}).refine((data) => {
  if (data.n_consensus > 1 && data.temperature === 0) {
    throw new Error('n_consensus greater than 1 but temperature is 0');
  }
  return true;
}).transform((data) => {
  // Handle document/documents compatibility
  if (data.documents && data.documents.length > 0) {
    return { ...data, document: data.documents[0] };
  } else if (data.document) {
    return { ...data, documents: [data.document] };
  }
  throw new Error('document or documents must be provided');
});
export type DocumentExtractRequest = z.infer<typeof DocumentExtractRequestSchema>;

// Consensus Model
export const ConsensusModelSchema = z.object({
  model: z.string(),
  temperature: z.number().default(0.0),
  reasoning_effort: ChatCompletionReasoningEffortSchema.default('medium'),
});
export type ConsensusModel = z.infer<typeof ConsensusModelSchema>;

// Field Location
export const FieldLocationSchema = z.object({
  label: z.string(),
  value: z.string(),
  quote: z.string(),
  file_id: z.string().nullable().optional(),
  page: z.number().int().nullable().optional(),
  bbox_normalized: z.tuple([z.number(), z.number(), z.number(), z.number()]).nullable().optional(),
  score: z.number().nullable().optional(),
  match_level: z.enum(['token', 'line', 'block']).nullable().optional(),
});
export type FieldLocation = z.infer<typeof FieldLocationSchema>;

// Parsed Choice
export const RetabParsedChoiceSchema = z.object({
  finish_reason: z.enum(['stop', 'length', 'tool_calls', 'content_filter', 'function_call']).nullable().optional(),
  field_locations: z.record(FieldLocationSchema).nullable().optional(),
  key_mapping: z.record(z.string().nullable()).nullable().optional(),
  message: z.object({
    content: z.string().nullable().optional(),
    role: z.string(),
    parsed: z.any().nullable().optional(),
  }),
  index: z.number().int(),
});
export type RetabParsedChoice = z.infer<typeof RetabParsedChoiceSchema>;

// Likelihoods source
export const LikelihoodsSourceSchema = z.enum(['consensus', 'log_probs']);
export type LikelihoodsSource = z.infer<typeof LikelihoodsSourceSchema>;

// Amount (for cost calculation)
export const AmountSchema = z.object({
  value: z.number(),
  currency: z.string(),
});
export type Amount = z.infer<typeof AmountSchema>;

// Usage stats
export const UsageSchema = z.object({
  prompt_tokens: z.number().int().optional(),
  completion_tokens: z.number().int().optional(),
  total_tokens: z.number().int().optional(),
});
export type Usage = z.infer<typeof UsageSchema>;

// Parsed Chat Completion
export const RetabParsedChatCompletionSchema = z.object({
  id: z.string(),
  object: z.literal('chat.completion'),
  created: z.number().int(),
  model: z.string(),
  choices: z.array(RetabParsedChoiceSchema),
  usage: UsageSchema.optional(),
  extraction_id: z.string().nullable().optional(),
  likelihoods: z.record(z.any()).nullable().optional(),
  schema_validation_error: z.custom<ErrorDetail>().nullable().optional(),
  request_at: z.string().datetime().nullable().optional(),
  first_token_at: z.string().datetime().nullable().optional(),
  last_token_at: z.string().datetime().nullable().optional(),
}).transform((data) => ({
  ...data,
  get api_cost(): Amount | null {
    // Implementation would compute cost from model and usage
    return null;
  },
}));
export type RetabParsedChatCompletion = z.infer<typeof RetabParsedChatCompletionSchema>;

// UI Response
export const UiResponseSchema = z.object({
  id: z.string(),
  object: z.string(),
  choices: z.array(z.any()),
  model: z.string(),
  extraction_id: z.string().nullable().optional(),
  likelihoods: z.record(z.any()).nullable().optional(),
  schema_validation_error: z.custom<ErrorDetail>().nullable().optional(),
  request_at: z.string().datetime().nullable().optional(),
  first_token_at: z.string().datetime().nullable().optional(),
  last_token_at: z.string().datetime().nullable().optional(),
});
export type UiResponse = z.infer<typeof UiResponseSchema>;

// Log Extraction Request
export const LogExtractionRequestSchema = z.object({
  messages: z.array(ChatCompletionRetabMessageSchema).nullable().optional(),
  openai_messages: z.array(z.any()).nullable().optional(),
  openai_responses_input: z.array(z.any()).nullable().optional(),
  anthropic_messages: z.array(z.any()).nullable().optional(),
  anthropic_system_prompt: z.string().nullable().optional(),
  document: MIMEDataSchema.default({
    id: 'dummy_doc',
    extension: 'txt',
    content: Buffer.from('No document provided').toString('base64'),
    mime_type: 'text/plain',
    unique_filename: 'dummy.txt',
    size: 20,
    filename: 'dummy.txt',
    url: 'data:text/plain;base64,' + Buffer.from('No document provided').toString('base64'),
  }),
  completion: z.union([z.record(z.any()), RetabParsedChatCompletionSchema]).nullable().optional(),
  openai_responses_output: z.any().nullable().optional(),
  json_schema: z.record(z.any()),
  model: z.string(),
  temperature: z.number(),
}).refine((data) => {
  const messagesCandidates = [
    data.messages,
    data.openai_messages,
    data.anthropic_messages,
    data.openai_responses_input,
  ].filter(candidate => candidate !== null && candidate !== undefined);
  
  if (messagesCandidates.length !== 1) {
    throw new Error('Exactly one of messages, openai_messages, anthropic_messages, openai_responses_input must be provided');
  }

  if (data.anthropic_messages && !data.anthropic_system_prompt) {
    throw new Error('anthropic_system_prompt must be provided if anthropic_messages is provided');
  }

  const completionCandidates = [
    data.completion,
    data.openai_responses_output,
  ].filter(candidate => candidate !== null && candidate !== undefined);
  
  if (completionCandidates.length !== 1) {
    throw new Error('Exactly one of completion, openai_responses_output must be provided');
  }

  return true;
});
export type LogExtractionRequest = z.infer<typeof LogExtractionRequestSchema>;

// Log Extraction Response
export const LogExtractionResponseSchema = z.object({
  extraction_id: z.string().nullable().optional(),
  status: z.enum(['success', 'error']),
  error_message: z.string().nullable().optional(),
});
export type LogExtractionResponse = z.infer<typeof LogExtractionResponseSchema>;

// Streaming types
export const RetabParsedChoiceDeltaChunkSchema = z.object({
  content: z.string().nullable().optional(),
  flat_likelihoods: z.record(z.number()).default({}),
  flat_parsed: z.record(z.any()).default({}),
  flat_deleted_keys: z.array(z.string()).default([]),
  field_locations: z.record(z.array(FieldLocationSchema)).nullable().optional(),
  is_valid_json: z.boolean().default(false),
  key_mapping: z.record(z.string().nullable()).nullable().optional(),
});
export type RetabParsedChoiceDeltaChunk = z.infer<typeof RetabParsedChoiceDeltaChunkSchema>;

export const RetabParsedChoiceChunkSchema = z.object({
  delta: RetabParsedChoiceDeltaChunkSchema,
  index: z.number().int(),
  finish_reason: z.enum(['stop', 'length', 'tool_calls', 'content_filter', 'function_call']).nullable().optional(),
});
export type RetabParsedChoiceChunk = z.infer<typeof RetabParsedChoiceChunkSchema>;

export const RetabParsedChatCompletionChunkSchema = z.object({
  id: z.string(),
  object: z.literal('chat.completion.chunk'),
  created: z.number().int(),
  model: z.string(),
  choices: z.array(RetabParsedChoiceChunkSchema),
  usage: UsageSchema.optional(),
  extraction_id: z.string().nullable().optional(),
  schema_validation_error: z.custom<ErrorDetail>().nullable().optional(),
  request_at: z.string().datetime().nullable().optional(),
  first_token_at: z.string().datetime().nullable().optional(),
  last_token_at: z.string().datetime().nullable().optional(),
  streaming_error: z.custom<ErrorDetail>().nullable().optional(),
}).transform((data) => ({
  ...data,
  get api_cost(): Amount | null {
    // Implementation would compute cost from model and usage
    return null;
  },
  chunk_accumulator(_previous?: RetabParsedChatCompletionChunk): RetabParsedChatCompletionChunk {
    // Implementation would accumulate chunks
    return data as any;
  },
}));
export type RetabParsedChatCompletionChunk = z.infer<typeof RetabParsedChatCompletionChunkSchema>;